### PR TITLE
Contained level support in teacher dashboard

### PR DESCRIPTION
**Pull request by balderdash**
Show progress and text responses for levels with contained levels in teacher dashboard.

When a student completes a level with a contained level, a new UserLevel is created for the contained level. This means whenever we render progress (or list text responses) we have to check if the level has a contained level, and if so check for a UserLevel associated with that.

I stacked these changes on top of https://github.com/code-dot-org/code-dot-org/pull/12624. Once I merge that, I'll change this PR's base to staging.

### CODE REVIEW

File: dashboard/config/scripts/levels/U3L03 - draw step.level [hunk 1]
```diff
@@ -66,9 +66,7 @@
     "encrypted_examples": "zkDI0HDKHo5c8J74y03IKljJJN88FpvWCVjolNR18sIy18c5dx0tu25snUyX\nRtR7\n",
     "lock_zero_param_functions": "true",
     "execute_palette_apis_only": "true",
-    "contained_level_names": [
-      "Draw Diamond"
-    ],
+    "contained_level_names": null,
```
Comment (joshlory) [inline dashboard/config/scripts/levels/U3L03 - draw step.level:69]: Intentional?

Comment (balderdash) [reply dashboard/config/scripts/levels/U3L03 - draw step.level:69]: Yeah, these levels don't exist. `Level#contained_levels` used to silently ignore them, but I think it's better to have it complain loudly.

File: dashboard/app/models/script.rb [hunk 2]
```diff
@@ -295,6 +295,36 @@ def self.get_from_cache(id)
...
+          text_response_levels << {
+            script_level: script_level,
+            levels: [level.contained_levels.first]
+          }
+        end
+      end
+    end
+
+    text_response_levels.concat(script_levels.includes(:levels).
+        where('levels.type' => [TextMatch, FreeResponse]).
```
Comment (joshlory) [inline dashboard/app/models/script.rb:317]: Why also `TextMatch`?

File: dashboard/app/models/level.rb [hunk 3]
```diff
@@ -338,7 +338,9 @@ def icon
   def contained_levels
     names = properties["contained_level_names"]
     return [] unless names.present?
-    Level.where(name: properties["contained_level_names"])
+    properties["contained_level_names"].map do |contained_level_name|
+      Script.cache_find_level(contained_level_name)
```
Comment (joshlory) [inline dashboard/app/models/level.rb:342]: 👍  changing this to use the cache.

File: dashboard/app/models/script.rb (hunk 2 above)
Comment (joshlory) [reply dashboard/app/models/script.rb:317]: Ah I see this is existing code moved from above.  Maybe cleaner if these both inherited from a class that exposed `last_attempt` as `last_answer_text`, but no need to change this.

**Pull request merged**